src/envoy/http/service_control/client_cache_test.cc

```diff
@@ -0,0 +1,179 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/envoy/http/service_control/client_cache.h"
+#include "common/common/empty_string.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "test/mocks/common.h"
+#include "test/mocks/event/mocks.h"
+#include "test/mocks/server/mocks.h"
+#include "test/mocks/stats/mocks.h"
+#include "test/test_common/utility.h"
+
+namespace espv2 {
+namespace envoy {
+namespace http_filters {
+namespace service_control {
+namespace {
+
+using ::espv2::api_proxy::service_control::CheckResponseInfo;
+using ::google::api::envoy::http::service_control::FilterConfig;
+using ::google::api::envoy::http::service_control::Service;
+using ::google::api::servicecontrol::v1::CheckError;
+using ::google::protobuf::util::Status;
+using ::google::protobuf::util::error::Code;
+using ::testing::NiceMock;
+
+class ClientCacheCheckResponseTest : public ::testing::Test {
+ protected:
+  ClientCacheCheckResponseTest() : stats_base_("test", context_.scope_) {
+    token_fn_ = []() { return Envoy::EMPTY_STRING; };
+  }
+
+  void SetUp() override {
+    cache_ = std::make_unique<ClientCache>(
+        service_config_, filter_config_, stats_base_.stats(), cm_, time_source_,
+        dispatcher_, token_fn_, token_fn_);
+  }
+
+  void CheckAndReset(Envoy::Stats::Counter& counter, const int expected_value) {
+    EXPECT_EQ(counter.value(), expected_value);
+    counter.reset();
+  }
+
+  void TearDown() override {
+    CheckAndReset(stats_base_.stats().filter_.allowed_control_plane_fault_, 0);
+    CheckAndReset(stats_base_.stats().filter_.denied_control_plane_fault_, 0);
+  }
+
+  // Helpers for SetUp.
+  Service service_config_;
+  FilterConfig filter_config_;
+  NiceMock<Envoy::Upstream::MockClusterManager> cm_;
+  NiceMock<Envoy::Event::MockDispatcher> dispatcher_;
+  NiceMock<Envoy::MockTimeSystem> time_source_;
+  NiceMock<Envoy::Server::Configuration::MockFactoryContext> context_;
+  ServiceControlFilterStatBase stats_base_;
+  std::function<const std::string&()> token_fn_;
+
+  // Class under test.
+  std::unique_ptr<ClientCache> cache_;
+};
+
+TEST_F(ClientCacheCheckResponseTest, Http5xxAllowed) {
+  CheckDoneFunc on_done = [](const Status& status, const CheckResponseInfo&) {
+    EXPECT_EQ(status.code(), Code::OK);
+  };
+
+  CheckResponse* resp = new CheckResponse();
+  const Status http_status(Code::UNAVAILABLE, "");
+  cache_->handleCheckResponse(http_status, resp, on_done);
+
+  CheckAndReset(stats_base_.stats().filter_.allowed_control_plane_fault_, 1);
```

Comment:
The main part of test can be put into a test separate function to be shared, which can even make it to be table-driven. Input with responseStatus, expectStatus and statsCounter.

Reply:
Same for ClientCacheCheckResponseNetworkFailClosedTest.

Reply:
I feel this is unnecessary, it will only save a few lines of code. And CheckResponse still needs to be constructed manually in each test.

Keep in mind table driven tests are go style, not c++ style. This is not recommended by gtest. We should be using test fixture parameters instead for gtest (which we do not in our codebase...)

Reply:
I think we are using test fixture parameters. 

It still helps to make code look clearer if we can extract the helper function. You can make `new response` in SetUp.

Reply:
Done, thanks it's clearer now.

I did not move the `new response` to `SetUp`, I feel this hides what is being tested. Each test is clearer when it creates the response and passes it to the test function.